### DISCUSSION TITLE
Build Milvus in Debug mode

### DISCUSSION THREAD

**Discussion by ashkrisk** [milvus-io/milvus#35116] [Q&A and General discussion] [answered]
Is there a way to create a debug build for Milvus? The Makefile doesn't seem to have a `debug` target.
The closest equivalent I can find is
```shell
make mode=debug
```
which should build the C++ code in debug mode, but the Go code is still built with all optimizations enabled.

I can manually edit the Makefile to disable optimizations for Go code as well, but that's a bit cumbersome. Also this means that I can't have both Debug and Release binaries at the same time.

Is there any other official way to create a debug build?

**Comment by yhmo** [comment, marked as answer]
So far as I know, we don't have an official way to build debug mode for Golang.
In the Makefile, we can see "go build" is executed here:
```
milvus: build-cpp print-build-info
	@echo "Building Milvus ..."
	@source $(PWD)/scripts/setenv.sh && \
		mkdir -p $(INSTALL_PATH) && go env -w CGO_ENABLED="1" && \
		GO111MODULE=on $(GO) build -pgo=$(PGO_PATH)/default.pgo -ldflags="-r $${RPATH} -X '$(OBJPREFIX).BuildTags=$(BUILD_TAGS)' -X '$(OBJPREFIX).BuildTime=$(BUILD_TIME)' -X '$(OBJPREFIX).GitCommit=$(GIT_COMMIT)' -X '$(OBJPREFIX).GoVersion=$(GO_VERSION)'" \
		-tags dynamic -o $(INSTALL_PATH)/milvus $(PWD)/cmd/main.go 1>/dev/null
```

We can manually add -gcflags="-N -l" to disable optimization and keep debug info:

```
milvus: build-cpp print-build-info
	@echo "Building Milvus ..."
	@source $(PWD)/scripts/setenv.sh && \
		mkdir -p $(INSTALL_PATH) && go env -w CGO_ENABLED="1" && \
		GO111MODULE=on $(GO) build -gcflags="-N -l" -pgo=$(PGO_PATH)/default.pgo -ldflags="-r $${RPATH} -X '$(OBJPREFIX).BuildTags=$(BUILD_TAGS)' -X '$(OBJPREFIX).BuildTime=$(BUILD_TIME)' -X '$(OBJPREFIX).GitCommit=$(GIT_COMMIT)' -X '$(OBJPREFIX).GoVersion=$(GO_VERSION)'" \
		-tags dynamic -o $(INSTALL_PATH)/milvus $(PWD)/cmd/main.go 1>/dev/null
```

Usually, milvus developers don't need to attach/debug milvus because there are lots of threads/gorutines in runtime, it might run into a weird state if you block one of the threads/gorutines.

**Reply by ashkrisk** [reply]
This almost worked for me, but I had to change `-gcflags="-N -l"` to `-gcflags="all=-N -l"`
